dimod/reference/samplers/exact_solver.py

```diff
@@ -86,46 +84,18 @@ def sample(self, bqm):
             :obj:`~dimod.SampleSet`
 
         """
-        M = bqm.binary.to_numpy_matrix()
-        off = bqm.binary.offset
-
-        if M.shape == (0, 0):
+        n = len(bqm)
+        if n == 0:
             return SampleSet.from_samples([], bqm.vartype, energy=[])
-
-        sample = np.zeros((len(bqm),), dtype=bool)
-
-        # now we iterate, flipping one bit at a time until we have
-        # traversed all samples. This is a Gray code.
-        # https://en.wikipedia.org/wiki/Gray_code
-        def iter_samples():
-            sample = np.zeros((len(bqm)), dtype=bool)
-            energy = 0.0
-
-            yield sample.copy(), energy + off
-
-            for i in range(1, 1 << len(bqm)):
-                v = _ffs(i)
-
-                # flip the bit in the sample
-                sample[v] = not sample[v]
-
-                # for now just calculate the energy, but there is a more clever way by calculating
-                # the energy delta for the single bit flip, don't have time, pull requests
-                # appreciated!
-                energy = sample.dot(M).dot(sample.transpose())
-
-                yield sample.copy(), float(energy) + off
-
-        samples, energies = zip(*iter_samples())
-
-        response = SampleSet.from_samples(np.array(samples, dtype='int8'), Vartype.BINARY, energies)
-
-        # make sure the response matches the given vartype, in-place.
+        bqm_copy = bqm.copy()
+        bqm_copy = bqm_copy.change_vartype(Vartype.BINARY)
+        samples = list(itertools.product([0, 1], repeat=n))
+        response = SampleSet.from_samples_bqm(samples, bqm_copy)
```

Comment:
Will bqms with non-sequential-integer variable names be handled?

Reply:
Do we have test cases for the non-sequential-integer variable? If yes, it passes, otherwise, give me an example to test and add.

Reply:
It won't, unless you do
```
response = SampleSet.from_samples_bqm((samples, list(bqm_copy)), bqm_copy)
```

Reply:
```
def test_arbitrary_labels(self):
    bqm = dimod.BQM.from_ising({}, {'ab': -1}
    sampleset = dimod.ExactSolver().sample(bqm)
    self.assertEqual(set(sampleset.variables), set(bqm))
```

Reply:
This test is passed.

Reply:
Ah, because if you send in a spin-valued BQM and it returns a binary sampleset it will fail

Reply:
Oh, I see why it's passing, because it's using the `bqm_index_labels` decorator. In the interests of readability and performance, I would remove the decorator and use the `response = SampleSet.from_samples_bqm((samples, list(bqm_copy)), bqm_copy)` syntax